File: client-sdks/python/sdk/model/client_option.py
# ------------------------------------------------------------------------------
# Copyright 2025 to Present, Jason Lam - VertexCache (https://github.com/vertexcache/vertexcache)
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ------------------------------------------------------------------------------

from sdk.model.encryption_mode import EncryptionMode


class ClientOption:
    DEFAULT_CLIENT_ID = "sdk-client"
    DEFAULT_HOST = "127.0.0.1"
    DEFAULT_PORT = 50505
    DEFAULT_READ_TIMEOUT = 3000
    DEFAULT_CONNECT_TIMEOUT = 3000

    def __init__(self):
        self.client_id: str = self.DEFAULT_CLIENT_ID
        self.client_token: str = ""

        self.server_host: str = self.DEFAULT_HOST
        self.server_port: int = self.DEFAULT_PORT

        self.enable_tls_encryption: bool = False
        self.tls_certificate: str = ""
        self.verify_certificate: bool = False

        self.encryption_mode: EncryptionMode = EncryptionMode.NONE
        self.encrypt_with_public_key: bool = False
        self.encrypt_with_shared_key: bool = False

        self.public_key: str = ""
        self.shared_encryption_key: str = ""

        self.read_timeout: int = self.DEFAULT_READ_TIMEOUT
        self.connect_timeout: int = self.DEFAULT_CONNECT_TIMEOUT

    def get_client_id(self) -> str:
        return self.client_id if self.client_id is not None else ""

    def get_client_token(self) -> str:
        return self.client_token if self.client_token is not None else ""

    def build_ident_command(self) -> str:
        return f'IDENT {{"client_id":"{self.get_client_id()}", "token":"{self.get_client_token()}"}}'

    def set_public_key(self, public_key: str):
        self.public_key = public_key

    def set_shared_encryption_key(self, shared_key: str):
        self.shared_encryption_key = shared_key

File: client-sdks/go/sdk/model/client_option.go
// ------------------------------------------------------------------------------
// Copyright 2025 to Present, Jason Lam - VertexCache (https://github.com/vertexcache)
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ------------------------------------------------------------------------------

package model

import (
	"fmt"
)

// ClientOption is a configuration container for initializing the VertexCache SDK client.
//
// This struct holds all user-specified options required to establish a connection
// to a VertexCache server, including host, port, TLS settings, authentication tokens,
// encryption modes (asymmetric or symmetric), and related keys or certificates.
//
// It provides a flexible way to customize client behavior, including security preferences.
type ClientOption struct {
	ClientID    string
	ClientToken string

	ServerHost string
	ServerPort int

	EnableTLSEncryption bool
	TLSCertificate      string
	VerifyCertificate   bool

	EncryptionMode       EncryptionMode
	EncryptWithPublicKey bool
	EncryptWithSharedKey bool

	PublicKey           string
	SharedEncryptionKey string

	ReadTimeout    int
	ConnectTimeout int
}

// NewClientOption returns a new ClientOption with default values set.
func NewClientOption() *ClientOption {
	return &ClientOption{
		ClientID:       "sdk-client",
		ServerHost:     "127.0.0.1",
		ServerPort:     50505,
		ReadTimeout:    3000,
		ConnectTimeout: 3000,
		EncryptionMode: None,
	}
}

// BuildIdentCommand constructs the IDENT command used during client handshake.
func (opt *ClientOption) BuildIdentCommand() string {
	clientID := opt.ClientID
	if clientID == "" {
		clientID = ""
	}
	token := opt.ClientToken
	if token == "" {
		token = ""
	}
	return fmt.Sprintf("IDENT {\"client_id\":\"%s\", \"token\":\"%s\"}", clientID, token)
}
